runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPortablePipelineTranslator.java

```diff
@@ -192,18 +204,28 @@ public StreamExecutionEnvironment getExecutionEnvironment() {
     translatorMap.put(PTransformTranslation.FLATTEN_TRANSFORM_URN, this::translateFlatten);
     translatorMap.put(PTransformTranslation.GROUP_BY_KEY_TRANSFORM_URN, this::translateGroupByKey);
     translatorMap.put(PTransformTranslation.IMPULSE_TRANSFORM_URN, this::translateImpulse);
-    translatorMap.put(STREAMING_IMPULSE_TRANSFORM_URN, this::translateStreamingImpulse);
     translatorMap.put(
         PTransformTranslation.ASSIGN_WINDOWS_TRANSFORM_URN, this::translateAssignWindows);
     translatorMap.put(ExecutableStage.URN, this::translateExecutableStage);
     translatorMap.put(PTransformTranslation.RESHUFFLE_URN, this::translateReshuffle);
 
+    // TODO Legacy transforms which need to be removed
+    // Consider removing now that timers are supported
+    translatorMap.put(STREAMING_IMPULSE_TRANSFORM_URN, this::translateStreamingImpulse);
+    // Remove once Reads can be wrapped in SDFs
+    translatorMap.put(PTransformTranslation.READ_TRANSFORM_URN, this::translateRead);
+
     this.urnToTransformTranslator = translatorMap.build();
   }
 
   @Override
   public Set<String> knownUrns() {
-    return urnToTransformTranslator.keySet();
+    // Do not expose Read as a known URN because we only want to support Read
```

Comment:
Why this special case? Probably related to the change in GreedyPipelineFuser?

Reply:
Cross language transforms can be more than just read so giving read special treatment might not be very useful.

Reply:
Read directly from Python or a cross-language read for Java from a Python pipeline will map to the same native read transform, right ? So not sure why we'd want to explicitly disallow this. (also true for any other native transform, for example, GBK)

Reply:
In portable pipelines Reads are executed via an environment, i.e. in the SDK harness via ExecutableStage. So they are not native. If we expose the Read URN here, we will have to deal with translating Python and Java Reads natively. However, we can only translate Java Reads natively.

Besides the code here being an optimization to execute without the Java SDK harness, we also want to support unbounded reads which are not supported via the SDK harness.